{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# TOOLS installing and testing\n",
    "\n",
    "Code for installing all necessary tools"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "#FASTQC\n",
    "\n",
    "#Seen in BIT09\n",
    "#change version to latest\n",
    "#REMARK: need terminal to add sudo pwd!\n",
    "\n",
    "cd /opt/\n",
    "sudo wget \"https://www.bioinformatics.babraham.ac.uk/projects/fastqc/fastqc_v0.11.9.zip\"\n",
    "sudo unzip fastqc_v0.11.9.zip\n",
    "sudo chmod 755 FastQC/fastqc\n",
    "sudo ln -s /opt/FastQC/fastqc /usr/local/bin/fastqc\n",
    "sudo rm fastqc_v0.11.9.zip\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "#MULTIQC\n",
    "\n",
    "#Seen in BIT09\n",
    "\n",
    "pip3 install --user multiqc\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "#Trimmomatic\n",
    "\n",
    "#Seen in BIT09\n",
    "#change version to latest\n",
    "#REMARK: need terminal to add sudo pwd!\n",
    "\n",
    "cd /opt/\n",
    "sudo wget \"http://www.usadellab.org/cms/uploads/supplementary/Trimmomatic/Trimmomatic-0.39.zip\"\n",
    "sudo unzip Trimmomatic-0.39.zip\n",
    "sudo chmod 755 Trimmomatic-0.39/trimmomatic-0.39.jar\n",
    "sudo rm Trimmomatic-0.39.zip\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "TrimmomaticPE: Started with arguments:\n",
      " -threads 2 -phred33 ERR1795072_1.fastq.gz ERR1795072_2.fastq.gz outputTrimmo/read1_paired.fastq.gz outputTrimmo/read1_unpaired.fastq.gz outputTrimmo/read2_paired.fastq.gz outputTrimmo/read2_unpaired.fastq.gz ILLUMINACLIP:/opt/Trimmomatic-0.39/adapters/NexteraPE-PE.fa:2:30:10 LEADING:3 TRAILING:3 SLIDINGWINDOW:4:15 MINLEN:35\n",
      "Using PrefixPair: 'AGATGTGTATAAGAGACAG' and 'AGATGTGTATAAGAGACAG'\n",
      "Using Long Clipping Sequence: 'GTCTCGTGGGCTCGGAGATGTGTATAAGAGACAG'\n",
      "Using Long Clipping Sequence: 'TCGTCGGCAGCGTCAGATGTGTATAAGAGACAG'\n",
      "Using Long Clipping Sequence: 'CTGTCTCTTATACACATCTCCGAGCCCACGAGAC'\n",
      "Using Long Clipping Sequence: 'CTGTCTCTTATACACATCTGACGCTGCCGACGA'\n",
      "ILLUMINACLIP: Using 1 prefix pairs, 4 forward/reverse sequences, 0 forward only sequences, 0 reverse only sequences\n",
      "Input Read Pairs: 1361900 Both Surviving: 1330331 (97,68%) Forward Only Surviving: 29107 (2,14%) Reverse Only Surviving: 1742 (0,13%) Dropped: 720 (0,05%)\n",
      "TrimmomaticPE: Completed successfully\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "#trimmomatic test\n",
    "\n",
    "cd ~/PROJECTHH/Scripts/ExampleInput/SamplesFastq\n",
    "mkdir outputTrimmo\n",
    "java -jar /opt/Trimmomatic-0.39/trimmomatic-0.39.jar PE -threads 2 -phred33 ERR1795072_1.fastq.gz ERR1795072_2.fastq.gz outputTrimmo/read1_paired.fastq.gz outputTrimmo/read1_unpaired.fastq.gz outputTrimmo/read2_paired.fastq.gz outputTrimmo/read2_unpaired.fastq.gz ILLUMINACLIP:/opt/Trimmomatic-0.39/adapters/NexteraPE-PE.fa:2:30:10 LEADING:3 TRAILING:3 SLIDINGWINDOW:4:15 MINLEN:35\n",
    "\n",
    "#Result: take a longer time than fastp (remark: also included more parameters)\n",
    "\n",
    "#SPLIT UP\n",
    "#java -jar /opt/Trimmomatic-0.39/trimmomatic-0.39.jar PE -threads 2 -phred33 ERR1795072_1.fastq.gz ERR1795072_2.fastq.gz \n",
    "#outputTrimmo/read1_paired.fastq.gz outputTrimmo/read1_unpaired.fastq.gz \n",
    "#outputTrimmo/read2_paired.fastq.gz outputTrimmo/read2_unpaired.fastq.gz \n",
    "#ILLUMINACLIP:/opt/Trimmomatic-0.39/adapters/NexteraPE-PE.fa:2:30:10 LEADING:3 TRAILING:3 SLIDINGWINDOW:4:15 MINLEN:35"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting package metadata (current_repodata.json): ...working... done\n",
      "Solving environment: ...working... done\n",
      "\n",
      "## Package Plan ##\n",
      "\n",
      "  environment location: /home/hannelore/miniconda3\n",
      "\n",
      "  added / updated specs:\n",
      "    - fastp\n",
      "\n",
      "\n",
      "The following packages will be downloaded:\n",
      "\n",
      "    package                    |            build\n",
      "    ---------------------------|-----------------\n",
      "    certifi-2019.11.28         |   py37hc8dfbb8_1         149 KB  conda-forge\n",
      "    conda-4.8.3                |   py37hc8dfbb8_1         3.0 MB  conda-forge\n",
      "    fastp-0.20.0               |       hdbcaa40_0         2.3 MB  bioconda\n",
      "    openssl-1.1.1e             |       h516909a_0         2.1 MB  conda-forge\n",
      "    python_abi-3.7             |          1_cp37m           4 KB  conda-forge\n",
      "    ------------------------------------------------------------\n",
      "                                           Total:         7.7 MB\n",
      "\n",
      "The following NEW packages will be INSTALLED:\n",
      "\n",
      "  fastp              bioconda/linux-64::fastp-0.20.0-hdbcaa40_0\n",
      "  python_abi         conda-forge/linux-64::python_abi-3.7-1_cp37m\n",
      "\n",
      "The following packages will be UPDATED:\n",
      "\n",
      "  certifi                                 2019.11.28-py37_0 --> 2019.11.28-py37hc8dfbb8_1\n",
      "  conda                                        4.8.0-py37_1 --> 4.8.3-py37hc8dfbb8_1\n",
      "  openssl                                 1.1.1d-h516909a_0 --> 1.1.1e-h516909a_0\n",
      "\n",
      "\n",
      "Proceed ([y]/n)? \n",
      "\n",
      "Downloading and Extracting Packages\n",
      "\r",
      "conda-4.8.3          | 3.0 MB    |            |   0% \r",
      "conda-4.8.3          | 3.0 MB    |            |   1% \r",
      "conda-4.8.3          | 3.0 MB    | 4          |   4% \r",
      "conda-4.8.3          | 3.0 MB    | 7          |   8% \r",
      "conda-4.8.3          | 3.0 MB    | #1         |  12% \r",
      "conda-4.8.3          | 3.0 MB    | #6         |  16% \r",
      "conda-4.8.3          | 3.0 MB    | ##         |  20% \r",
      "conda-4.8.3          | 3.0 MB    | ##4        |  24% \r",
      "conda-4.8.3          | 3.0 MB    | ##8        |  28% \r",
      "conda-4.8.3          | 3.0 MB    | ###2       |  33% \r",
      "conda-4.8.3          | 3.0 MB    | ###6       |  37% \r",
      "conda-4.8.3          | 3.0 MB    | ####       |  40% \r",
      "conda-4.8.3          | 3.0 MB    | ####4      |  44% \r",
      "conda-4.8.3          | 3.0 MB    | ####8      |  49% \r",
      "conda-4.8.3          | 3.0 MB    | #####2     |  53% \r",
      "conda-4.8.3          | 3.0 MB    | #####6     |  57% \r",
      "conda-4.8.3          | 3.0 MB    | ######     |  61% \r",
      "conda-4.8.3          | 3.0 MB    | ######5    |  65% \r",
      "conda-4.8.3          | 3.0 MB    | ######9    |  69% \r",
      "conda-4.8.3          | 3.0 MB    | #######3   |  73% \r",
      "conda-4.8.3          | 3.0 MB    | #######7   |  77% \r",
      "conda-4.8.3          | 3.0 MB    | ########1  |  82% \r",
      "conda-4.8.3          | 3.0 MB    | ########5  |  86% \r",
      "conda-4.8.3          | 3.0 MB    | ########9  |  90% \r",
      "conda-4.8.3          | 3.0 MB    | #########4 |  94% \r",
      "conda-4.8.3          | 3.0 MB    | #########8 |  98% \r",
      "conda-4.8.3          | 3.0 MB    | ########## | 100% \n",
      "\r",
      "fastp-0.20.0         | 2.3 MB    |            |   0% \r",
      "fastp-0.20.0         | 2.3 MB    |            |   1% \r",
      "fastp-0.20.0         | 2.3 MB    | 5          |   6% \r",
      "fastp-0.20.0         | 2.3 MB    | #1         |  11% \r",
      "fastp-0.20.0         | 2.3 MB    | #5         |  16% \r",
      "fastp-0.20.0         | 2.3 MB    | ##1        |  21% \r",
      "fastp-0.20.0         | 2.3 MB    | ##6        |  27% \r",
      "fastp-0.20.0         | 2.3 MB    | ###1       |  32% \r",
      "fastp-0.20.0         | 2.3 MB    | ###6       |  37% \r",
      "fastp-0.20.0         | 2.3 MB    | ####1      |  42% \r",
      "fastp-0.20.0         | 2.3 MB    | ####6      |  47% \r",
      "fastp-0.20.0         | 2.3 MB    | #####1     |  52% \r",
      "fastp-0.20.0         | 2.3 MB    | #####7     |  57% \r",
      "fastp-0.20.0         | 2.3 MB    | ######2    |  63% \r",
      "fastp-0.20.0         | 2.3 MB    | ######7    |  68% \r",
      "fastp-0.20.0         | 2.3 MB    | #######3   |  73% \r",
      "fastp-0.20.0         | 2.3 MB    | #######8   |  78% \r",
      "fastp-0.20.0         | 2.3 MB    | ########4  |  84% \r",
      "fastp-0.20.0         | 2.3 MB    | ########9  |  90% \r",
      "fastp-0.20.0         | 2.3 MB    | #########5 |  95% \r",
      "fastp-0.20.0         | 2.3 MB    | ########## | 100% \n",
      "\r",
      "python_abi-3.7       | 4 KB      |            |   0% \r",
      "python_abi-3.7       | 4 KB      | ########## | 100% \n",
      "\r",
      "openssl-1.1.1e       | 2.1 MB    |            |   0% \r",
      "openssl-1.1.1e       | 2.1 MB    |            |   1% \r",
      "openssl-1.1.1e       | 2.1 MB    | 6          |   7% \r",
      "openssl-1.1.1e       | 2.1 MB    | #2         |  12% \r",
      "openssl-1.1.1e       | 2.1 MB    | #8         |  18% \r",
      "openssl-1.1.1e       | 2.1 MB    | ##4        |  24% \r",
      "openssl-1.1.1e       | 2.1 MB    | ##9        |  30% \r",
      "openssl-1.1.1e       | 2.1 MB    | ###5       |  36% \r",
      "openssl-1.1.1e       | 2.1 MB    | ####       |  41% \r",
      "openssl-1.1.1e       | 2.1 MB    | ####7      |  48% \r",
      "openssl-1.1.1e       | 2.1 MB    | #####3     |  53% \r",
      "openssl-1.1.1e       | 2.1 MB    | #####9     |  59% \r",
      "openssl-1.1.1e       | 2.1 MB    | ######5    |  66% \r",
      "openssl-1.1.1e       | 2.1 MB    | #######1   |  72% \r",
      "openssl-1.1.1e       | 2.1 MB    | #######7   |  78% \r",
      "openssl-1.1.1e       | 2.1 MB    | ########3  |  83% \r",
      "openssl-1.1.1e       | 2.1 MB    | ########9  |  89% \r",
      "openssl-1.1.1e       | 2.1 MB    | #########5 |  95% \r",
      "openssl-1.1.1e       | 2.1 MB    | ########## | 100% \n",
      "\r",
      "certifi-2019.11.28   | 149 KB    |            |   0% \r",
      "certifi-2019.11.28   | 149 KB    | ####3      |  43% \r",
      "certifi-2019.11.28   | 149 KB    | ########## | 100% \n",
      "Preparing transaction: ...working... done\n",
      "Verifying transaction: ...working... done\n",
      "Executing transaction: ...working... done\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "#fastp\n",
    "\n",
    "conda install -c bioconda fastp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Read1 before filtering:\n",
      "total reads: 1361900\n",
      "total bases: 323356824\n",
      "Q20 bases: 309068137(95.5811%)\n",
      "Q30 bases: 300820639(93.0306%)\n",
      "\n",
      "Read2 before filtering:\n",
      "total reads: 1361900\n",
      "total bases: 323723854\n",
      "Q20 bases: 289437255(89.4087%)\n",
      "Q30 bases: 272544633(84.1905%)\n",
      "\n",
      "Read1 after filtering:\n",
      "total reads: 1313860\n",
      "total bases: 311533482\n",
      "Q20 bases: 299607900(96.172%)\n",
      "Q30 bases: 292374974(93.8503%)\n",
      "\n",
      "Read2 aftering filtering:\n",
      "total reads: 1313860\n",
      "total bases: 311522132\n",
      "Q20 bases: 283738741(91.0814%)\n",
      "Q30 bases: 268872443(86.3093%)\n",
      "\n",
      "Filtering result:\n",
      "reads passed filter: 2627720\n",
      "reads failed due to low quality: 96080\n",
      "reads failed due to too many N: 0\n",
      "reads failed due to too short: 0\n",
      "reads with adapter trimmed: 18928\n",
      "bases trimmed due to adapters: 364273\n",
      "\n",
      "Duplication rate: 0.995976%\n",
      "\n",
      "Insert size peak (evaluated by paired-end reads): 328\n",
      "\n",
      "JSON report: fastp.json\n",
      "HTML report: fastp.html\n",
      "\n",
      "fastp -i ERR1795072_1.fastq.gz -I ERR1795072_2.fastq.gz -o outputFastp/out.R1.fq.gz -O outputFastp/out.R2.fq.gz \n",
      "fastp v0.20.0, time used: 26 seconds\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "#fastp test\n",
    "\n",
    "cd ~/PROJECTHH/Scripts/ExampleInput/SamplesFastq\n",
    "mkdir outputFastp\n",
    "fastp -i ERR1795072_1.fastq.gz -I ERR1795072_2.fastq.gz -o outputFastp/out.R1.fq.gz -O outputFastp/out.R2.fq.gz\n",
    "\n",
    "#Å–ESULT: fast analysis but no extra parametesr included + nice overview in html\n",
    "# remark:but low filtering percentage (reads were probably already filtered) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "#MiniKraken (for taxonomic sequence classification)\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
