\documentclass[12pt,a4paper]{article}
\begin{document}
\begin{table}[ht]
\begin{center}
\caption{All statistics are based on contigs of size $\geq$ 500 bp, unless otherwise noted (e.g., "\# contigs ($\geq$ 0 bp)" and "Total length ($\geq$ 0 bp)" include all contigs).}
\begin{tabular}{|l*{17}{|r}|}
\hline
Assembly & TRIM\_ERR1795072\_Illumina\_MiSeq\_paired\_end\_sequencing\_EXP\_0\_1\_2017\_01\_13\_14\_47\_56\_mini\_1.contigs & TRIM\_ERR1795073\_Illumina\_MiSeq\_paired\_end\_sequencing\_EXP\_1\_1\_2017\_01\_13\_14\_47\_56\_mini\_1.contigs & TRIM\_ERR1795074\_Illumina\_MiSeq\_paired\_end\_sequencing\_EXP\_2\_1\_2017\_01\_13\_14\_47\_56\_mini\_1.contigs & TRIM\_ERR1795075\_Illumina\_MiSeq\_paired\_end\_sequencing\_EXP\_3\_1\_2017\_01\_13\_14\_47\_56\_mini\_1.contigs & TRIM\_ERR1795076\_Illumina\_MiSeq\_paired\_end\_sequencing\_EXP\_4\_1\_2017\_01\_13\_14\_47\_56\_mini\_1.contigs & TRIM\_ERR1795077\_Illumina\_MiSeq\_paired\_end\_sequencing\_EXP\_5\_1\_2017\_01\_13\_14\_47\_56\_mini\_1.contigs & TRIM\_ERR1795078\_Illumina\_MiSeq\_paired\_end\_sequencing\_EXP\_6\_1\_2017\_01\_13\_14\_47\_56\_mini\_1.contigs & TRIM\_ERR1795079\_Illumina\_MiSeq\_paired\_end\_sequencing\_EXP\_7\_1\_2017\_01\_13\_14\_47\_56\_mini\_1.contigs & TRIM\_ERR1795080\_Illumina\_MiSeq\_paired\_end\_sequencing\_EXP\_8\_1\_2017\_01\_13\_14\_47\_56\_mini\_1.contigs & TRIM\_ERR1795081\_Illumina\_MiSeq\_paired\_end\_sequencing\_EXP\_9\_1\_2017\_01\_13\_14\_47\_56\_mini\_1.contigs & TRIM\_ERR1795082\_Illumina\_MiSeq\_paired\_end\_sequencing\_EXP\_10\_1\_2017\_01\_13\_14\_47\_56\_mini\_1.contigs & TRIM\_ERR1795083\_Illumina\_MiSeq\_paired\_end\_sequencing\_EXP\_11\_1\_2017\_01\_13\_14\_47\_56\_mini\_1.contigs & TRIM\_ERR1795084\_Illumina\_MiSeq\_paired\_end\_sequencing\_EXP\_12\_1\_2017\_01\_13\_14\_47\_56\_mini\_1.contigs & TRIM\_ERR1795085\_Illumina\_MiSeq\_paired\_end\_sequencing\_EXP\_13\_1\_2017\_01\_13\_14\_47\_56\_mini\_1.contigs & TRIM\_ERR1795086\_Illumina\_MiSeq\_paired\_end\_sequencing\_EXP\_14\_1\_2017\_01\_13\_14\_47\_56\_mini\_1.contigs & TRIM\_ERR1795087\_Illumina\_MiSeq\_paired\_end\_sequencing\_EXP\_15\_1\_2017\_01\_13\_14\_47\_56\_mini\_1.contigs & TRIM\_SO\_9245\_KP\_mini\_1.contigs \\ \hline
\# contigs ($\geq$ 1000 bp) & 16 & 13 & 20 & 12 & 33 & 20 & 85 & 1 & 9 & 11 & 15 & 9 & 14 & 15 & 11 & 5 & 2 \\ \hline
\# contigs ($\geq$ 5000 bp) & 0 & 0 & 0 & 0 & 0 & 0 & 1 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 \\ \hline
\# contigs ($\geq$ 10000 bp) & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 \\ \hline
\# contigs ($\geq$ 25000 bp) & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 \\ \hline
\# contigs ($\geq$ 50000 bp) & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 \\ \hline
Total length ($\geq$ 1000 bp) & 23274 & 20862 & 24832 & 16390 & 50815 & 25173 & {\bf 172034} & 1665 & 13186 & 14774 & 19167 & 12804 & 19199 & 20445 & 16122 & 7466 & 3330 \\ \hline
Total length ($\geq$ 5000 bp) & 0 & 0 & 0 & 0 & 0 & 0 & {\bf 8044} & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 \\ \hline
Total length ($\geq$ 10000 bp) & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 \\ \hline
Total length ($\geq$ 25000 bp) & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 \\ \hline
Total length ($\geq$ 50000 bp) & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 \\ \hline
\# contigs & 183 & 157 & 192 & 173 & 214 & 176 & 203 & 119 & 146 & 174 & 156 & 141 & 183 & 179 & 182 & 145 & 47 \\ \hline
Largest contig & 2468 & 3720 & 1833 & 1901 & 2813 & 2220 & {\bf 8044} & 1665 & 2433 & 1726 & 2073 & 1985 & 2001 & 2242 & 2950 & 2150 & 1758 \\ \hline
Total length & 127509 & 112071 & 130797 & 116975 & 166002 & 124494 & {\bf 250774} & 72850 & 99625 & 115890 & 108519 & 94152 & 124394 & 119504 & 119769 & 91996 & 29758 \\ \hline
Reference length & 5724448 & 5724448 & 5724448 & 5724448 & 5724448 & 5724448 & 5724448 & 5724448 & 5724448 & 5724448 & 5724448 & 5724448 & 5724448 & 5724448 & 5724448 & 5724448 & 5724448 \\ \hline
GC (\%) & 54.45 & 55.11 & 54.14 & 56.22 & 56.02 & 55.46 & 53.56 & 55.34 & 55.71 & 56.20 & 54.89 & 56.46 & 55.32 & 55.39 & 56.22 & 56.53 & 59.70 \\ \hline
Reference GC (\%) & 57.22 & 57.22 & 57.22 & 57.22 & 57.22 & 57.22 & 57.22 & 57.22 & 57.22 & 57.22 & 57.22 & 57.22 & 57.22 & 57.22 & 57.22 & 57.22 & 57.22 \\ \hline
N50 & 645 & 659 & 659 & 638 & 743 & 684 & {\bf 1567} & 588 & 643 & 620 & 664 & 622 & 626 & 612 & 615 & 592 & 595 \\ \hline
N75 & 556 & 563 & 547 & 554 & 573 & 568 & {\bf 838} & 529 & 560 & 555 & 558 & 557 & 551 & 543 & 547 & 544 & 539 \\ \hline
L50 & 69 & 59 & 74 & 69 & 72 & 67 & 47 & 51 & 58 & 70 & 60 & 56 & 71 & 70 & 73 & 61 & {\bf 20} \\ \hline
L75 & 123 & 105 & 130 & 118 & 137 & 118 & 103 & 84 & 99 & 119 & 105 & 97 & 124 & 122 & 125 & 101 & {\bf 33} \\ \hline
\# misassemblies & 12 & 8 & 11 & 8 & 10 & 8 & 16 & {\bf 4} & 5 & {\bf 4} & 7 & 9 & 9 & 10 & 13 & {\bf 4} & {\bf 4} \\ \hline
\# misassembled contigs & 12 & 8 & 11 & 6 & 10 & 7 & 15 & 4 & 5 & 4 & 7 & 8 & 9 & 10 & 12 & 4 & {\bf 3} \\ \hline
Misassembled contigs length & 13257 & 6867 & 12859 & 7297 & 11944 & 9208 & 47127 & 3774 & 4950 & 3385 & 7223 & 7675 & 10758 & 11415 & 12513 & 3816 & {\bf 2826} \\ \hline
\# local misassemblies & {\bf 0} & 3 & {\bf 0} & {\bf 0} & {\bf 0} & {\bf 0} & 2 & 1 & 2 & 1 & 1 & {\bf 0} & {\bf 0} & {\bf 0} & 1 & {\bf 0} & {\bf 0} \\ \hline
\# scaffold gap ext. mis. & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 \\ \hline
\# scaffold gap loc. mis. & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 \\ \hline
\# unaligned mis. contigs & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 \\ \hline
\# unaligned contigs & 1 + 0 part & 0 + 2 part & {\bf 0 + 0 part} & 0 + 1 part & {\bf 0 + 0 part} & 1 + 0 part & 0 + 1 part & {\bf 0 + 0 part} & 0 + 3 part & {\bf 0 + 0 part} & {\bf 0 + 0 part} & 0 + 1 part & {\bf 0 + 0 part} & {\bf 0 + 0 part} & {\bf 0 + 0 part} & {\bf 0 + 0 part} & {\bf 0 + 0 part} \\ \hline
Unaligned length & 693 & 2175 & {\bf 0} & 555 & {\bf 0} & 609 & 550 & {\bf 0} & 2450 & {\bf 0} & {\bf 0} & 1045 & {\bf 0} & {\bf 0} & {\bf 0} & {\bf 0} & {\bf 0} \\ \hline
Genome fraction (\%) & 2.187 & 1.881 & 2.266 & 2.021 & 2.866 & 2.138 & {\bf 4.313} & 1.243 & 1.660 & 2.011 & 1.881 & 1.594 & 2.154 & 2.068 & 2.078 & 1.597 & 0.512 \\ \hline
Duplication ratio & 1.013 & 1.021 & 1.009 & {\bf 1.006} & 1.012 & 1.012 & 1.014 & 1.023 & 1.023 & 1.007 & 1.008 & 1.020 & 1.009 & 1.010 & 1.007 & 1.007 & 1.015 \\ \hline
\# N's per 100 kbp & 0.00 & 0.00 & 0.00 & 0.00 & 0.00 & 0.00 & 0.00 & 0.00 & 0.00 & 0.00 & 0.00 & 0.00 & 0.00 & 0.00 & 0.00 & 0.00 & 0.00 \\ \hline
\# mismatches per 100 kbp & 277.16 & 499.75 & 215.13 & 330.11 & 218.85 & 290.11 & 194.02 & 453.79 & 276.79 & {\bf 183.28} & 233.05 & 347.34 & 230.30 & 213.73 & 263.18 & 228.68 & 624.17 \\ \hline
\# indels per 100 kbp & 14.38 & 13.93 & 12.34 & 10.37 & 12.19 & 17.16 & 13.77 & 5.62 & 13.68 & 7.82 & 5.57 & 7.67 & 21.89 & 28.72 & {\bf 4.20} & 13.13 & 6.82 \\ \hline
Largest alignment & 2468 & 3719 & 1382 & 1901 & 2721 & 2018 & {\bf 4665} & 973 & 2190 & 1726 & 1938 & 1985 & 1729 & 1915 & 2950 & 2150 & 1758 \\ \hline
Total aligned length & 125362 & 107832 & 129748 & 115727 & 164240 & 122397 & {\bf 247902} & 71176 & 95059 & 115146 & 107777 & 91263 & 123390 & 118374 & 119015 & 91406 & 29305 \\ \hline
NA50 & 631 & 624 & 620 & 614 & 708 & 662 & {\bf 1341} & 579 & 619 & 611 & 652 & 593 & 614 & 602 & 597 & 579 & 587 \\ \hline
NGA50 & - & - & - & - & - & - & - & - & - & - & - & - & - & - & - & - & - \\ \hline
NA75 & 544 & 549 & 537 & 548 & 555 & 552 & {\bf 781} & 523 & 536 & 543 & 551 & 537 & 542 & 533 & 539 & 539 & 523 \\ \hline
LA50 & 74 & 63 & 80 & 72 & 76 & 70 & 57 & 54 & 63 & 72 & 64 & 62 & 76 & 75 & 77 & 62 & {\bf 21} \\ \hline
LA75 & 129 & 111 & 137 & 122 & 143 & 122 & 118 & 87 & 106 & 122 & 110 & 103 & 130 & 128 & 130 & 103 & {\bf 35} \\ \hline
\end{tabular}
\end{center}
\end{table}
\end{document}
