\documentclass[12pt,a4paper]{article}
\begin{document}
\begin{table}[ht]
\begin{center}
\caption{All statistics are based on contigs of size $\geq$ 500 bp, unless otherwise noted (e.g., "\# contigs ($\geq$ 0 bp)" and "Total length ($\geq$ 0 bp)" include all contigs).}
\begin{tabular}{|l*{17}{|r}|}
\hline
Assembly & TRIM\_ERR1795072\_Illumina\_MiSeq\_paired\_end\_sequencing\_EXP\_0\_1\_2017\_01\_13\_14\_47\_56\_mini\_1.contigs & TRIM\_ERR1795073\_Illumina\_MiSeq\_paired\_end\_sequencing\_EXP\_1\_1\_2017\_01\_13\_14\_47\_56\_mini\_1.contigs & TRIM\_ERR1795074\_Illumina\_MiSeq\_paired\_end\_sequencing\_EXP\_2\_1\_2017\_01\_13\_14\_47\_56\_mini\_1.contigs & TRIM\_ERR1795075\_Illumina\_MiSeq\_paired\_end\_sequencing\_EXP\_3\_1\_2017\_01\_13\_14\_47\_56\_mini\_1.contigs & TRIM\_ERR1795076\_Illumina\_MiSeq\_paired\_end\_sequencing\_EXP\_4\_1\_2017\_01\_13\_14\_47\_56\_mini\_1.contigs & TRIM\_ERR1795077\_Illumina\_MiSeq\_paired\_end\_sequencing\_EXP\_5\_1\_2017\_01\_13\_14\_47\_56\_mini\_1.contigs & TRIM\_ERR1795078\_Illumina\_MiSeq\_paired\_end\_sequencing\_EXP\_6\_1\_2017\_01\_13\_14\_47\_56\_mini\_1.contigs & TRIM\_ERR1795079\_Illumina\_MiSeq\_paired\_end\_sequencing\_EXP\_7\_1\_2017\_01\_13\_14\_47\_56\_mini\_1.contigs & TRIM\_ERR1795080\_Illumina\_MiSeq\_paired\_end\_sequencing\_EXP\_8\_1\_2017\_01\_13\_14\_47\_56\_mini\_1.contigs & TRIM\_ERR1795081\_Illumina\_MiSeq\_paired\_end\_sequencing\_EXP\_9\_1\_2017\_01\_13\_14\_47\_56\_mini\_1.contigs & TRIM\_ERR1795082\_Illumina\_MiSeq\_paired\_end\_sequencing\_EXP\_10\_1\_2017\_01\_13\_14\_47\_56\_mini\_1.contigs & TRIM\_ERR1795083\_Illumina\_MiSeq\_paired\_end\_sequencing\_EXP\_11\_1\_2017\_01\_13\_14\_47\_56\_mini\_1.contigs & TRIM\_ERR1795084\_Illumina\_MiSeq\_paired\_end\_sequencing\_EXP\_12\_1\_2017\_01\_13\_14\_47\_56\_mini\_1.contigs & TRIM\_ERR1795085\_Illumina\_MiSeq\_paired\_end\_sequencing\_EXP\_13\_1\_2017\_01\_13\_14\_47\_56\_mini\_1.contigs & TRIM\_ERR1795086\_Illumina\_MiSeq\_paired\_end\_sequencing\_EXP\_14\_1\_2017\_01\_13\_14\_47\_56\_mini\_1.contigs & TRIM\_ERR1795087\_Illumina\_MiSeq\_paired\_end\_sequencing\_EXP\_15\_1\_2017\_01\_13\_14\_47\_56\_mini\_1.contigs & TRIM\_SO\_9245\_KP\_mini\_1.contigs \\ \hline
\# contigs ($\geq$ 1000 bp) & 3 & 4 & 5 & 4 & 9 & 7 & 4 & 5 & 6 & 5 & 8 & 8 & 7 & 4 & 7 & 5 & 2 \\ \hline
\# contigs ($\geq$ 5000 bp) & 0 & 0 & 0 & 1 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 \\ \hline
\# contigs ($\geq$ 10000 bp) & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 \\ \hline
\# contigs ($\geq$ 25000 bp) & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 \\ \hline
\# contigs ($\geq$ 50000 bp) & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 \\ \hline
Total length ($\geq$ 1000 bp) & 6849 & 5789 & 7223 & 9839 & 11183 & 9677 & 6291 & 10194 & 8640 & 6675 & 11982 & {\bf 12886} & 11848 & 6355 & 12541 & 8418 & 3330 \\ \hline
Total length ($\geq$ 5000 bp) & 0 & 0 & 0 & {\bf 5887} & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 \\ \hline
Total length ($\geq$ 10000 bp) & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 \\ \hline
Total length ($\geq$ 25000 bp) & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 \\ \hline
Total length ($\geq$ 50000 bp) & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 \\ \hline
\# contigs & 78 & 73 & 77 & 85 & 89 & 77 & 36 & 90 & 97 & 106 & 74 & 102 & 85 & 101 & 105 & 117 & 50 \\ \hline
Largest contig & 3915 & 2222 & 1925 & {\bf 5887} & 1722 & 2220 & 2079 & 3218 & 2190 & 1759 & 2102 & 2384 & 3663 & 2242 & 3553 & 2150 & 1758 \\ \hline
Total length & 52799 & 48185 & 50168 & 59050 & 60567 & 51282 & 25486 & 61237 & 64386 & 69808 & 53284 & 70095 & 59963 & 63756 & 70794 & {\bf 74900} & 31718 \\ \hline
Reference length & 5577703 & 5577703 & 5577703 & 5577703 & 5577703 & 5577703 & 5577703 & 5577703 & 5577703 & 5577703 & 5577703 & 5577703 & 5577703 & 5577703 & 5577703 & 5577703 & 5577703 \\ \hline
GC (\%) & 56.62 & 56.30 & 55.62 & 57.20 & 57.82 & 56.68 & 55.16 & 55.64 & 55.61 & 57.41 & 55.79 & 56.78 & 56.73 & 56.72 & 57.12 & 56.76 & 59.54 \\ \hline
Reference GC (\%) & 57.26 & 57.26 & 57.26 & 57.26 & 57.26 & 57.26 & 57.26 & 57.26 & 57.26 & 57.26 & 57.26 & 57.26 & 57.26 & 57.26 & 57.26 & 57.26 & 57.26 \\ \hline
N50 & 591 & 628 & 583 & 608 & 630 & 618 & 623 & 595 & 627 & 619 & {\bf 668} & 624 & 625 & 590 & 609 & 587 & 599 \\ \hline
N75 & 532 & 553 & 540 & 544 & 551 & 536 & 539 & 528 & 555 & 546 & {\bf 562} & 553 & 549 & 524 & 542 & 543 & 539 \\ \hline
L50 & 30 & 30 & 31 & 31 & 35 & 30 & {\bf 13} & 34 & 39 & 43 & 27 & 39 & 31 & 42 & 40 & 48 & 21 \\ \hline
L75 & 53 & 51 & 53 & 57 & 60 & 53 & {\bf 24} & 61 & 67 & 73 & 49 & 69 & 57 & 70 & 71 & 81 & 35 \\ \hline
\# misassemblies & 4 & 1 & 2 & 1 & 2 & 3 & {\bf 0} & 2 & 3 & 3 & 2 & 3 & 3 & 3 & 4 & 2 & 4 \\ \hline
\# misassembled contigs & 3 & 1 & 2 & 1 & 2 & 3 & {\bf 0} & 2 & 3 & 3 & 2 & 3 & 3 & 3 & 4 & 2 & 4 \\ \hline
Misassembled contigs length & 2553 & 709 & 1895 & 686 & 1451 & 2993 & {\bf 0} & 1527 & 2695 & 2523 & 2950 & 1900 & 2986 & 2234 & 4067 & 1357 & 3419 \\ \hline
\# local misassemblies & 1 & {\bf 0} & {\bf 0} & {\bf 0} & {\bf 0} & {\bf 0} & {\bf 0} & 1 & 1 & 2 & {\bf 0} & {\bf 0} & {\bf 0} & 1 & {\bf 0} & {\bf 0} & {\bf 0} \\ \hline
\# scaffold gap ext. mis. & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 \\ \hline
\# scaffold gap loc. mis. & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 \\ \hline
\# unaligned mis. contigs & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 \\ \hline
\# unaligned contigs & 0 + 1 part & 0 + 2 part & 0 + 1 part & {\bf 0 + 0 part} & {\bf 0 + 0 part} & {\bf 0 + 0 part} & 0 + 1 part & 0 + 3 part & 0 + 2 part & 0 + 1 part & 0 + 1 part & 0 + 1 part & 0 + 1 part & {\bf 0 + 0 part} & 0 + 1 part & 0 + 1 part & {\bf 0 + 0 part} \\ \hline
Unaligned length & 657 & 1440 & 637 & {\bf 0} & {\bf 0} & {\bf 0} & 873 & 2112 & 1189 & 1465 & 733 & 1686 & 1139 & {\bf 0} & 635 & 1031 & {\bf 0} \\ \hline
Genome fraction (\%) & 0.915 & 0.815 & 0.882 & 1.033 & 1.070 & 0.904 & 0.441 & 1.025 & 1.109 & 1.224 & 0.931 & 1.201 & 1.033 & 1.114 & 1.227 & {\bf 1.315} & 0.560 \\ \hline
Duplication ratio & 1.021 & 1.028 & 1.006 & 1.025 & 1.014 & 1.017 & {\bf 1.000} & 1.034 & 1.021 & 1.001 & 1.012 & 1.022 & 1.021 & 1.026 & 1.026 & 1.007 & 1.015 \\ \hline
\# N's per 100 kbp & 0.00 & 0.00 & 0.00 & 0.00 & 0.00 & 0.00 & 0.00 & 0.00 & 0.00 & 0.00 & 0.00 & 0.00 & 0.00 & 0.00 & 0.00 & 0.00 & 0.00 \\ \hline
\# mismatches per 100 kbp & 517.08 & 448.81 & 489.62 & 545.17 & 520.87 & 511.83 & 552.58 & 656.10 & 588.25 & 492.22 & 483.55 & 506.25 & 518.97 & 522.98 & 556.93 & {\bf 440.44} & 687.96 \\ \hline
\# indels per 100 kbp & 25.46 & 13.20 & 8.13 & 13.89 & 10.05 & 3.97 & 4.06 & 12.25 & 14.54 & 14.65 & 11.56 & 8.96 & 12.15 & 6.44 & {\bf 2.92} & 8.18 & 9.60 \\ \hline
Largest alignment & 3831 & 2222 & 1925 & {\bf 5540} & 1722 & 2105 & 1884 & 3218 & 2190 & 1759 & 2020 & 2268 & 3420 & 2104 & 3440 & 2150 & 1758 \\ \hline
Total aligned length & 51198 & 45442 & 49224 & 57601 & 59797 & 50407 & 24611 & 57439 & 62007 & 68260 & 51917 & 66960 & 57615 & 62478 & 68422 & {\bf 73333} & 31237 \\ \hline
NA50 & 580 & 595 & 574 & 601 & 624 & 589 & 604 & 578 & 606 & 608 & {\bf 643} & 593 & 587 & 577 & 587 & 577 & 595 \\ \hline
NGA50 & - & - & - & - & - & - & - & - & - & - & - & - & - & - & - & - & - \\ \hline
NA75 & 526 & 527 & 527 & 541 & 544 & 528 & 532 & 517 & 533 & 534 & {\bf 545} & 534 & 533 & 521 & 529 & 534 & 518 \\ \hline
LA50 & 32 & 32 & 33 & 34 & 36 & 31 & {\bf 15} & 38 & 42 & 46 & 30 & 43 & 35 & 43 & 44 & 50 & 22 \\ \hline
LA75 & 56 & 53 & 55 & 60 & 62 & 54 & {\bf 26} & 66 & 71 & 76 & 53 & 74 & 61 & 72 & 76 & 84 & 37 \\ \hline
\end{tabular}
\end{center}
\end{table}
\end{document}
