{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# MLST ANALYSIS\n",
    "## 15/11/2019 \n",
    "\n",
    "### Checking Chewbbaca to do MLST analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Doing the tutorial\n",
    "Ref: https://github.com/B-UMMI/chewBBACA_tutorial"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#BASH: sudo pip3 install chewbacca\n",
    "#BASH: git clone https://github.com/B-UMMI/chewBBACA_tutorial #get all the files to do the tutorial"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " \u001b[0m\u001b[38;5;33mchewBBACA_tutorial\u001b[0m                'GIT - projectHH.ipynb'          readme.txt\n",
      " \u001b[38;5;33mdataKP-assembly\u001b[0m                   'MLST-analysis (bash).ipynb'\n",
      "'Getting and prepping data.ipynb'  'MLST analysis (python).ipynb'\n",
      "\u001b[0m\u001b[38;5;33manalysis_all\u001b[0m  \u001b[38;5;33mgenes\u001b[0m    info.tsv   \u001b[38;5;33mresults_all\u001b[0m  Streptococcus_agalactiae.trn\n",
      "\u001b[38;5;33manalysis_cg\u001b[0m   \u001b[38;5;33mgenomes\u001b[0m  README.md  \u001b[38;5;33mresults_cg\u001b[0m\n"
     ]
    }
   ],
   "source": [
    "ls\n",
    "cd chewBBACA_tutorial\n",
    "ls"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Archive:  genomes/complete_genomes.zip\n",
      "   creating: complete_genomes/\n",
      "  inflating: complete_genomes/GCA_000007265.1_ASM726v1_genomic.fna  \n",
      "  inflating: complete_genomes/GCA_000012705.1_ASM1270v1_genomic.fna  \n",
      "  inflating: complete_genomes/GCA_000196055.1_ASM19605v1_genomic.fna  \n",
      "  inflating: complete_genomes/GCA_000299135.1_ASM29913v1_genomic.fna  \n",
      "  inflating: complete_genomes/GCA_000302475.2_ASM30247v2_genomic.fna  \n",
      "  inflating: complete_genomes/GCA_000427035.1_09mas018883_genomic.fna  \n",
      "  inflating: complete_genomes/GCA_000427055.1_ILRI112_genomic.fna  \n",
      "  inflating: complete_genomes/GCA_000427075.1_ILRI005_genomic.fna  \n",
      "  inflating: complete_genomes/GCA_000599965.1_ASM59996v1_genomic.fna  \n",
      "  inflating: complete_genomes/GCA_000636115.1_ASM63611v1_genomic.fna  \n",
      "  inflating: complete_genomes/GCA_000689235.1_GBCO_p1_genomic.fna  \n",
      "  inflating: complete_genomes/GCA_000730215.2_ASM73021v2_genomic.fna  \n",
      "  inflating: complete_genomes/GCA_000730255.1_ASM73025v1_genomic.fna  \n",
      "  inflating: complete_genomes/GCA_000782855.1_ASM78285v1_genomic.fna  \n",
      "  inflating: complete_genomes/GCA_000831105.1_ASM83110v1_genomic.fna  \n",
      "  inflating: complete_genomes/GCA_000831125.1_ASM83112v1_genomic.fna  \n",
      "  inflating: complete_genomes/GCA_000831145.1_ASM83114v1_genomic.fna  \n",
      "  inflating: complete_genomes/GCA_000967445.1_ASM96744v1_genomic.fna  \n",
      "  inflating: complete_genomes/GCA_001026925.1_ASM102692v1_genomic.fna  \n",
      "  inflating: complete_genomes/GCA_001190805.1_ASM119080v1_genomic.fna  \n",
      "  inflating: complete_genomes/GCA_001190825.1_ASM119082v1_genomic.fna  \n",
      "  inflating: complete_genomes/GCA_001190845.1_ASM119084v1_genomic.fna  \n",
      "  inflating: complete_genomes/GCA_001190865.1_ASM119086v1_genomic.fna  \n",
      "  inflating: complete_genomes/GCA_001190885.1_ASM119088v1_genomic.fna  \n",
      "  inflating: complete_genomes/GCA_001266635.1_ASM126663v1_genomic.fna  \n",
      "  inflating: complete_genomes/GCA_001275545.2_ASM127554v2_genomic.fna  \n",
      "  inflating: complete_genomes/GCA_001448985.1_ASM144898v1_genomic.fna  \n",
      "  inflating: complete_genomes/GCA_001552035.1_ASM155203v1_genomic.fna  \n",
      "  inflating: complete_genomes/GCA_001592385.1_ASM159238v1_genomic.fna  \n",
      "  inflating: complete_genomes/GCA_001592425.1_ASM159242v1_genomic.fna  \n",
      "  inflating: complete_genomes/GCA_001655175.1_ASM165517v1_genomic.fna  \n",
      "  inflating: complete_genomes/GCA_001683515.1_ASM168351v1_genomic.fna  \n"
     ]
    }
   ],
   "source": [
    "unzip genomes/complete_genomes.zip"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#BASH: chewBBACA.py CreateSchema -i complete_genomes/ --cpu 4 -o schema_seed --ptf Streptococcus_agalactiae.trn \n",
    "#error missing prodigal files\n",
    "\n",
    "#Check if is prodigal installed? No -> install it:\n",
    "#BASH: sudo pip3 install prodigal #does not work\n",
    "#BASH: sudo dnf install prodigal #does not work\n",
    "#BASH: conda install -c bioconda prodigal #found package: installing"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 22/11/2019 : continue the tutorial"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "totaal 2940\n",
      "drwxrwxr-x.  4 hannelore hannelore    4096 15 nov 16:12 \u001b[0m\u001b[38;5;33manalysis_all\u001b[0m\n",
      "drwxrwxr-x.  2 hannelore hannelore    4096 15 nov 16:12 \u001b[38;5;33manalysis_cg\u001b[0m\n",
      "drwxrwxr-x.  2 hannelore hannelore    4096 15 mei  2017 \u001b[38;5;33mcomplete_genomes\u001b[0m\n",
      "drwxrwxr-x.  2 hannelore hannelore    4096 15 nov 16:12 \u001b[38;5;33mgenes\u001b[0m\n",
      "drwxrwxr-x.  2 hannelore hannelore    4096 15 nov 16:12 \u001b[38;5;33mgenomes\u001b[0m\n",
      "-rw-rw-r--.  1 hannelore hannelore   33630 15 nov 16:12 info.tsv\n",
      "-rw-rw-r--.  1 hannelore hannelore    3235 22 nov 16:59 listGenomes2Call.txt\n",
      "-rw-rw-r--.  1 hannelore hannelore 2868848 22 nov 17:00 proteinID_Genome.tsv\n",
      "-rw-rw-r--.  1 hannelore hannelore   10954 15 nov 16:12 README.md\n",
      "drwxrwxr-x.  3 hannelore hannelore    4096 15 nov 16:12 \u001b[38;5;33mresults_all\u001b[0m\n",
      "drwxrwxr-x.  3 hannelore hannelore    4096 15 nov 16:12 \u001b[38;5;33mresults_cg\u001b[0m\n",
      "-rw-rw-r--.  1 hannelore hannelore  558392 15 nov 16:12 Streptococcus_agalactiae.trn\n",
      "drwxrwxr-x. 33 hannelore hannelore    4096 15 nov 16:37 \u001b[38;5;33mtemp\u001b[0m\n"
     ]
    }
   ],
   "source": [
    "#retry:\n",
    "# BASH: chewBBACA.py CreateSchema -i complete_genomes/ --cpu 4 -o schema_seed --ptf Streptococcus_agalactiae.trn\n",
    "\n",
    "ll ~/PROJECTHH/chewBBACA_tutorial\n",
    "#result: proteinID_Genome.tsv BUT No folder schema_seed (??)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### output:\n",
    "(base) [hannelore@localhost chewBBACA_tutorial]$ chewBBACA.py CreateSchema -i complete_genomes/ --cpu 4 -o schema_seed --ptf Streptococcus_agalactiae.trn\n",
    "chewBBACA version 2.0.17.2 by Mickael Silva at https://github.com/B-UMMI/chewBBACA\n",
    "email contact: mickaelsilva@medicina.ulisboa.pt\n",
    "\n",
    "\n",
    "Warning, you are close to use all your cpus, if you are using a laptop you may be uncapable to perform any action\n",
    "will use this training file : /home/hannelore/PROJECTHH/chewBBACA_tutorial/Streptococcus_agalactiae.trn\n",
    "Will use this number of cpus: 4\n",
    "Checking all programs are installed\n",
    "Checking Prodigal installed... True\n",
    "\n",
    "Starting Script at : 16:59:34-22/11/2019\n",
    "\n",
    "Starting Prodigal at : 16:59:34-22/11/2019\n",
    "chosen taxon :/home/hannelore/PROJECTHH/chewBBACA_tutorial/Streptococcus_agalactiae.trn\n",
    "done prodigal run on:GCA_000012705.1_ASM1270v1_genomic.fna\n",
    "done prodigal run on:GCA_000007265.1_ASM726v1_genomic.fna\n",
    "done prodigal run on:GCA_000299135.1_ASM29913v1_genomic.fna\n",
    "done prodigal run on:GCA_000196055.1_ASM19605v1_genomic.fna\n",
    "done prodigal run on:GCA_000302475.2_ASM30247v2_genomic.fna\n",
    "done prodigal run on:GCA_000427055.1_ILRI112_genomic.fna\n",
    "done prodigal run on:GCA_000427035.1_09mas018883_genomic.fna\n",
    "done prodigal run on:GCA_000427075.1_ILRI005_genomic.fna\n",
    "done prodigal run on:GCA_000599965.1_ASM59996v1_genomic.fna\n",
    "done prodigal run on:GCA_000636115.1_ASM63611v1_genomic.fna\n",
    "done prodigal run on:GCA_000689235.1_GBCO_p1_genomic.fna\n",
    "done prodigal run on:GCA_000730215.2_ASM73021v2_genomic.fna\n",
    "done prodigal run on:GCA_000730255.1_ASM73025v1_genomic.fna\n",
    "done prodigal run on:GCA_000782855.1_ASM78285v1_genomic.fna\n",
    "done prodigal run on:GCA_000831105.1_ASM83110v1_genomic.fna\n",
    "done prodigal run on:GCA_000831145.1_ASM83114v1_genomic.fna\n",
    "done prodigal run on:GCA_000967445.1_ASM96744v1_genomic.fna\n",
    "done prodigal run on:GCA_000831125.1_ASM83112v1_genomic.fna\n",
    "done prodigal run on:GCA_001190805.1_ASM119080v1_genomic.fna\n",
    "done prodigal run on:GCA_001026925.1_ASM102692v1_genomic.fna\n",
    "done prodigal run on:GCA_001190825.1_ASM119082v1_genomic.fna\n",
    "done prodigal run on:GCA_001190845.1_ASM119084v1_genomic.fna\n",
    "done prodigal run on:GCA_001190865.1_ASM119086v1_genomic.fna\n",
    "done prodigal run on:GCA_001266635.1_ASM126663v1_genomic.fna\n",
    "done prodigal run on:GCA_001190885.1_ASM119088v1_genomic.fna\n",
    "done prodigal run on:GCA_001275545.2_ASM127554v2_genomic.fna\n",
    "done prodigal run on:GCA_001448985.1_ASM144898v1_genomic.fna\n",
    "done prodigal run on:GCA_001552035.1_ASM155203v1_genomic.fna\n",
    "done prodigal run on:GCA_001592425.1_ASM159242v1_genomic.fna\n",
    "done prodigal run on:GCA_001655175.1_ASM165517v1_genomic.fna\n",
    "done prodigal run on:GCA_001592385.1_ASM159238v1_genomic.fna\n",
    "done prodigal run on:GCA_001683515.1_ASM168351v1_genomic.fna\n",
    "\n",
    "Checking all prodigal processes created the necessary files...\n",
    "All prodigal files necessary were created\n",
    "\n",
    "Finishing Prodigal at : 16:59:49-22/11/2019\n",
    "running analysis for pair : /home/hannelore/PROJECTHH/chewBBACA_tutorial/complete_genomes/GCA_000007265.1_ASM726v1_genomic.fna /home/hannelore/PROJECTHH/chewBBACA_tutorial/complete_genomes/GCA_000012705.1_ASM1270v1_genomic.fna\n",
    "running analysis for pair : /home/hannelore/PROJECTHH/chewBBACA_tutorial/complete_genomes/GCA_000196055.1_ASM19605v1_genomic.fna /home/hannelore/PROJECTHH/chewBBACA_tutorial/complete_genomes/GCA_000299135.1_ASM29913v1_genomic.fna\n",
    "running analysis for pair : /home/hannelore/PROJECTHH/chewBBACA_tutorial/complete_genomes/GCA_000302475.2_ASM30247v2_genomic.fna /home/hannelore/PROJECTHH/chewBBACA_tutorial/complete_genomes/GCA_000427035.1_09mas018883_genomic.fna\n",
    "running analysis for pair : /home/hannelore/PROJECTHH/chewBBACA_tutorial/complete_genomes/GCA_000427055.1_ILRI112_genomic.fna /home/hannelore/PROJECTHH/chewBBACA_tutorial/complete_genomes/GCA_000427075.1_ILRI005_genomic.fna\n",
    "running analysis for pair : /home/hannelore/PROJECTHH/chewBBACA_tutorial/complete_genomes/GCA_000599965.1_ASM59996v1_genomic.fna /home/hannelore/PROJECTHH/chewBBACA_tutorial/complete_genomes/GCA_000636115.1_ASM63611v1_genomic.fna\n",
    "running analysis for pair : /home/hannelore/PROJECTHH/chewBBACA_tutorial/complete_genomes/GCA_000689235.1_GBCO_p1_genomic.fna /home/hannelore/PROJECTHH/chewBBACA_tutorial/complete_genomes/GCA_000730215.2_ASM73021v2_genomic.fna\n",
    "running analysis for pair : /home/hannelore/PROJECTHH/chewBBACA_tutorial/complete_genomes/GCA_000730255.1_ASM73025v1_genomic.fna /home/hannelore/PROJECTHH/chewBBACA_tutorial/complete_genomes/GCA_000782855.1_ASM78285v1_genomic.fna\n",
    "running analysis for pair : /home/hannelore/PROJECTHH/chewBBACA_tutorial/complete_genomes/GCA_000831105.1_ASM83110v1_genomic.fna /home/hannelore/PROJECTHH/chewBBACA_tutorial/complete_genomes/GCA_000831125.1_ASM83112v1_genomic.fna\n",
    "running analysis for pair : /home/hannelore/PROJECTHH/chewBBACA_tutorial/complete_genomes/GCA_000831145.1_ASM83114v1_genomic.fna /home/hannelore/PROJECTHH/chewBBACA_tutorial/complete_genomes/GCA_000967445.1_ASM96744v1_genomic.fna\n",
    "running analysis for pair : /home/hannelore/PROJECTHH/chewBBACA_tutorial/complete_genomes/GCA_001026925.1_ASM102692v1_genomic.fna /home/hannelore/PROJECTHH/chewBBACA_tutorial/complete_genomes/GCA_001190805.1_ASM119080v1_genomic.fna\n",
    "running analysis for pair : /home/hannelore/PROJECTHH/chewBBACA_tutorial/complete_genomes/GCA_001190825.1_ASM119082v1_genomic.fna /home/hannelore/PROJECTHH/chewBBACA_tutorial/complete_genomes/GCA_001190845.1_ASM119084v1_genomic.fna\n",
    "running analysis for pair : /home/hannelore/PROJECTHH/chewBBACA_tutorial/complete_genomes/GCA_001190865.1_ASM119086v1_genomic.fna /home/hannelore/PROJECTHH/chewBBACA_tutorial/complete_genomes/GCA_001190885.1_ASM119088v1_genomic.fna\n",
    "running analysis for pair : /home/hannelore/PROJECTHH/chewBBACA_tutorial/complete_genomes/GCA_001266635.1_ASM126663v1_genomic.fna /home/hannelore/PROJECTHH/chewBBACA_tutorial/complete_genomes/GCA_001275545.2_ASM127554v2_genomic.fna\n",
    "running analysis for pair : /home/hannelore/PROJECTHH/chewBBACA_tutorial/complete_genomes/GCA_001448985.1_ASM144898v1_genomic.fna /home/hannelore/PROJECTHH/chewBBACA_tutorial/complete_genomes/GCA_001552035.1_ASM155203v1_genomic.fna\n",
    "running analysis for pair : /home/hannelore/PROJECTHH/chewBBACA_tutorial/complete_genomes/GCA_001592385.1_ASM159238v1_genomic.fna /home/hannelore/PROJECTHH/chewBBACA_tutorial/complete_genomes/GCA_001592425.1_ASM159242v1_genomic.fna\n",
    "running analysis for pair : /home/hannelore/PROJECTHH/chewBBACA_tutorial/complete_genomes/GCA_001655175.1_ASM165517v1_genomic.fna /home/hannelore/PROJECTHH/chewBBACA_tutorial/complete_genomes/GCA_001683515.1_ASM168351v1_genomic.fna\n",
    "sh: makeblastdb: opdracht niet gevonden\n",
    "sh: blastp: opdracht niet gevonden\n",
    "sh: makeblastdb: opdracht niet gevonden\n",
    "sh: blastp: opdracht niet gevonden\n",
    "sh: makeblastdb: opdracht niet gevonden\n",
    "sh: blastp: opdracht niet gevonden\n",
    "sh: makeblastdb: opdracht niet gevonden\n",
    "sh: blastp: opdracht niet gevonden\n",
    "sh: makeblastdb: opdracht niet gevonden\n",
    "sh: blastp: opdracht niet gevonden\n",
    "sh: makeblastdb: opdracht niet gevonden\n",
    "sh: blastp: opdracht niet gevonden\n",
    "sh: makeblastdb: opdracht niet gevonden\n",
    "sh: blastp: opdracht niet gevonden\n",
    "sh: makeblastdb: opdracht niet gevonden\n",
    "sh: blastp: opdracht niet gevonden\n",
    "sh: makeblastdb: opdracht niet gevonden\n",
    "sh: blastp: opdracht niet gevonden\n",
    "sh: makeblastdb: opdracht niet gevonden\n",
    "sh: blastp: opdracht niet gevonden\n",
    "sh: makeblastdb: opdracht niet gevonden\n",
    "sh: blastp: opdracht niet gevonden\n",
    "sh: makeblastdb: opdracht niet gevonden\n",
    "sh: blastp: opdracht niet gevonden\n",
    "sh: makeblastdb: opdracht niet gevonden\n",
    "sh: blastp: opdracht niet gevonden\n",
    "sh: makeblastdb: opdracht niet gevonden\n",
    "sh: blastp: opdracht niet gevonden\n",
    "sh: makeblastdb: opdracht niet gevonden\n",
    "sh: blastp: opdracht niet gevonden\n",
    "sh: makeblastdb: opdracht niet gevonden\n",
    "sh: blastp: opdracht niet gevonden\n",
    "running analysis for pair : temp/protogenome0/protogenome0.fasta temp/protogenome1/protogenome1.fasta\n",
    "running analysis for pair : temp/protogenome2/protogenome2.fasta temp/protogenome3/protogenome3.fasta\n",
    "running analysis for pair : temp/protogenome4/protogenome4.fasta temp/protogenome5/protogenome5.fasta\n",
    "running analysis for pair : temp/protogenome6/protogenome6.fasta temp/protogenome7/protogenome7.fasta\n",
    "running analysis for pair : temp/protogenome8/protogenome8.fasta temp/protogenome9/protogenome9.fasta\n",
    "running analysis for pair : temp/protogenome10/protogenome10.fasta temp/protogenome11/protogenome11.fasta\n",
    "running analysis for pair : temp/protogenome12/protogenome12.fasta temp/protogenome13/protogenome13.fasta\n",
    "running analysis for pair : temp/protogenome14/protogenome14.fasta temp/protogenome15/protogenome15.fasta\n",
    "sh: makeblastdb: opdracht niet gevonden\n",
    "sh: blastp: opdracht niet gevonden\n",
    "sh: makeblastdb: opdracht niet gevonden\n",
    "sh: blastp: opdracht niet gevonden\n",
    "sh: makeblastdb: opdracht niet gevonden\n",
    "sh: blastp: opdracht niet gevonden\n",
    "sh: makeblastdb: opdracht niet gevonden\n",
    "sh: blastp: opdracht niet gevonden\n",
    "sh: makeblastdb: opdracht niet gevonden\n",
    "sh: blastp: opdracht niet gevonden\n",
    "sh: makeblastdb: opdracht niet gevonden\n",
    "sh: blastp: opdracht niet gevonden\n",
    "sh: makeblastdb: opdracht niet gevonden\n",
    "sh: blastp: opdracht niet gevonden\n",
    "sh: makeblastdb: opdracht niet gevonden\n",
    "sh: blastp: opdracht niet gevonden\n",
    "running analysis for pair : temp/protogenome16/protogenome16.fasta temp/protogenome17/protogenome17.fasta\n",
    "running analysis for pair : temp/protogenome18/protogenome18.fasta temp/protogenome19/protogenome19.fasta\n",
    "running analysis for pair : temp/protogenome20/protogenome20.fasta temp/protogenome21/protogenome21.fasta\n",
    "running analysis for pair : temp/protogenome22/protogenome22.fasta temp/protogenome23/protogenome23.fasta\n",
    "sh: makeblastdb: opdracht niet gevonden\n",
    "sh: blastp: opdracht niet gevonden\n",
    "sh: makeblastdb: opdracht niet gevonden\n",
    "sh: blastp: opdracht niet gevonden\n",
    "sh: makeblastdb: opdracht niet gevonden\n",
    "sh: blastp: opdracht niet gevonden\n",
    "sh: makeblastdb: opdracht niet gevonden\n",
    "sh: blastp: opdracht niet gevonden\n",
    "running analysis for pair : temp/protogenome24/protogenome24.fasta temp/protogenome25/protogenome25.fasta\n",
    "running analysis for pair : temp/protogenome26/protogenome26.fasta temp/protogenome27/protogenome27.fasta\n",
    "sh: makeblastdb: opdracht niet gevonden\n",
    "sh: blastp: opdracht niet gevonden\n",
    "sh: makeblastdb: opdracht niet gevonden\n",
    "sh: blastp: opdracht niet gevonden\n",
    "running analysis for pair : temp/protogenome28/protogenome28.fasta temp/protogenome29/protogenome29.fasta\n",
    "sh: makeblastdb: opdracht niet gevonden\n",
    "sh: blastp: opdracht niet gevonden\n",
    "sh: makeblastdb: opdracht niet gevonden\n",
    "sh: blastp: opdracht niet gevonden\n",
    "[Errno 2] No such file or directory: '/home/hannelore/PROJECTHH/chewBBACA_tutorial/temp/protogenome30/proteins.xml'\n",
    "\n",
    "\tUSAGE : chewBBACA.py [module] -h \n",
    "\n",
    "Select one of the following functions :\n",
    "\n",
    "CreateSchema : Create a gene by gene schema based on genomes\n",
    "AlleleCall : Perform allele call for target genomes\n",
    "SchemaEvaluator : Tool that builds an html output to better navigate/visualize your schema\n",
    "TestGenomeQuality : Analyze your allele call output to refine schemas\n",
    "ExtractCgMLST : Select a subset of loci without missing data (to be used as PHYLOViZ input)\n",
    "RemoveGenes : Remove a provided list of loci from your allele call output\n",
    "PrepExternalSchema : prepare an external schema to be used by chewBBACA\n",
    "JoinProfiles : join two profiles in a single profile file\n",
    "UniprotFinder : get info about a schema created with chewBBACA\n",
    "\n",
    "\n",
    "\n",
    "###### maybe output folder was not found?\n",
    "##### work on 3 cpu in stead of 4 (4 is max on virtualbox)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Bash",
   "language": "bash",
   "name": "bash"
  },
  "language_info": {
   "codemirror_mode": "shell",
   "file_extension": ".sh",
   "mimetype": "text/x-sh",
   "name": "bash"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
